File: 25.k-个一组翻转链表.py
#
# @lc app=leetcode.cn id=25 lang=python3
#
# [25] K 个一组翻转链表
#
# https://leetcode-cn.com/problems/reverse-nodes-in-k-group/description/
#
# algorithms
# Hard (58.07%)
# Likes:    614
# Dislikes: 0
# Total Accepted:    77.2K
# Total Submissions: 125.5K
# Testcase Example:  '[1,2,3]\n2'
#
# 给你一个链表，每 k 个节点一组进行翻转，请你返回翻转后的链表。
#
# k 是一个正整数，它的值小于或等于链表的长度。
#
# 如果节点总数不是 k 的整数倍，那么请将最后剩余的节点保持原有顺序。
#
#
#
# 示例：
#
# 给你这个链表：1->2->3->4->5
#
# 当 k = 2 时，应当返回: 2->1->4->3->5
#
# 当 k = 3 时，应当返回: 3->2->1->4->5
#
#
#
# 说明：
#
#
# 你的算法只能使用常数的额外空间。
# 你不能只是单纯的改变节点内部的值，而是需要实际进行节点交换。
#
#
#

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
# @lc code=start
# Definition for singly-linked list.


class Solution:
    def reverseKGroup1(self, head: ListNode, k: int) -> ListNode:
        sentry = ListNode(None)
        pre_tail = sentry
        next_head = head

        while True:
            this_head = next_head
            for _ in range(k):
                if not next_head:
                    pre_tail.next = this_head
                    return sentry.next
                next_head = next_head.next

            pre = next_head
            cur = this_head
            for _ in range(k):
                cur.next, cur, pre = pre, cur.next, cur
            pre_tail.next = pre
            pre_tail = this_head

    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return head
        cur = head
        for _ in range(k-1):
            cur = cur.next
            if not cur:
                return head

        next_head = self.reverseKGroup(cur.next, k)
        pre = next_head
        cur = head
        for _ in range(k):
            cur.next, cur, pre = pre, cur.next, cur
        return pre

    # @lc code=end

File: 25.k-个一组翻转链表.go
/*
 * @lc app=leetcode.cn id=25 lang=golang
 *
 * [25] K 个一组翻转链表
 *
 * https://leetcode.cn/problems/reverse-nodes-in-k-group/description/
 *
 * algorithms
 * Hard (67.68%)
 * Likes:    2015
 * Dislikes: 0
 * Total Accepted:    458.9K
 * Total Submissions: 678K
 * Testcase Example:  '[1,2,3,4,5]\n2'
 *
 * 给你链表的头节点 head ，每 k 个节点一组进行翻转，请你返回修改后的链表。
 *
 * k 是一个正整数，它的值小于或等于链表的长度。如果节点总数不是 k 的整数倍，那么请将最后剩余的节点保持原有顺序。
 *
 * 你不能只是单纯的改变节点内部的值，而是需要实际进行节点交换。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：head = [1,2,3,4,5], k = 2
 * 输出：[2,1,4,3,5]
 *
 *
 * 示例 2：
 *
 *
 *
 *
 * 输入：head = [1,2,3,4,5], k = 3
 * 输出：[3,2,1,4,5]
 *
 *
 *
 * 提示：
 *
 *
 * 链表中的节点数目为 n
 * 1 <= k <= n <= 5000
 * 0 <= Node.val <= 1000
 *
 *
 *
 *
 * 进阶：你可以设计一个只用 O(1) 额外内存空间的算法解决此问题吗？
 *
 *
 *
 *
 */
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	sentry := &ListNode{Next: head}
	preTail := sentry
	lc := head

	for {
		rc := lc

		// check remain
		for i := 0; i < k; i++ {
			if rc == nil {
				preTail.Next = lc
				return sentry.Next
			}
			rc = rc.Next
		}

		// reverse
		var pre *ListNode
		cur := lc
		for cur != rc {
			cur.Next, pre, cur = pre, cur, cur.Next
		}
		preTail.Next = pre
		preTail = lc

		lc = rc
	}
	return sentry.Next
}

// @lc code=end
